tests/apps/test_sample.py

```diff
@@ -368,3 +369,48 @@ def test_vibronic_integration(p, integration_sample_number):
     assert dims == (integration_sample_number, len(alpha) * 2)
     assert samples.dtype == "int"
     assert (samples >= 0).all()
+
+
+class TestWawMatrix:
+    """Tests for the function ``strawberryfields.apps.sample.waw_matrix``"""
+
+    adjs = [
+        np.array(
+            [[0, 1, 0, 1, 1], [1, 0, 0, 1, 1], [0, 0, 0, 1, 0], [1, 1, 1, 0, 1], [1, 1, 0, 1, 0]]
+        ),
+        np.array([[0, 1, 0, 1], [1, 0, 1, 0], [0, 1, 0, 1], [1, 0, 1, 0]]),
+        np.array([[0, 1, 2], [1, 0, 0.5], [2, 0.5, 0]]),
+    ]
+
+    wvecs = [np.array([1, 1, 3, 1, 0.5]), np.array([1, -2, 3, -4]), np.array([0, 0.5, 4])]
+
+    resc_adjs = [
+        np.array(
+            [
+                [0, 1, 0, 1, 0.5],
+                [1, 0, 0, 1, 0.5],
+                [0, 0, 0, 3, 0],
+                [1, 1, 3, 0, 0.5],
+                [0.5, 0.5, 0, 0.5, 0],
+            ]
+        ),
+        np.array([[0, -2, 0, -4], [-2, 0, -6, 0], [0, -6, 0, -12], [-4, 0, -12, 0]]),
+        np.array([[0, 0, 0], [0, 0, 1], [0, 1, 0]]),
+    ]
+
+    def test_invalid_adjacency(self):
+        """Test if function raises a ``ValueError`` for a matrix that is not symmetric."""
+        with pytest.raises(ValueError, match="Input must be a NumPy array"):
+            adj_asym = np.triu(np.ones((4, 4)))
+            sample.sample(A=adj_asym, n_mean=1.0)
+
+    @pytest.mark.parametrize("inst", zip(adjs, wvecs, resc_adjs))
+    def test_valid_w(self, inst):
+        """Test if function returns the correct answer on some pre-calculated instances."""
+        assert np.allclose(sample.waw_matrix(inst[0], inst[1]), inst[2])
+
+    @pytest.mark.parametrize("inst", zip(adjs, wvecs, resc_adjs))
+    def test_valid_w_list(self, inst):
+        """Test if function returns the correct answer on some pre-calculated instances,
+        when w is a list"""
+        assert np.allclose(sample.waw_matrix(inst[0], list(inst[1])), inst[2])
```

Comment:
Nice!

Reply:
How were the matrices pre-calculated? By hand?

Reply:
Yes all by :hand:
Tried to cover a variety of cases (e.g. edges with different weights in `A`, positive/negative/fractional node weights)